src/java/org/apache/cassandra/index/sai/memory/MemtableIndex.java

```diff
@@ -19,39 +19,71 @@
 package org.apache.cassandra.index.sai.memory;
 
 import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
 import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
 import java.util.concurrent.atomic.LongAdder;
 
 import org.apache.cassandra.db.Clustering;
+import org.apache.cassandra.db.ClusteringComparator;
 import org.apache.cassandra.db.DecoratedKey;
 import org.apache.cassandra.db.PartitionPosition;
+import org.apache.cassandra.db.marshal.AbstractType;
+import org.apache.cassandra.db.memtable.Memtable;
+import org.apache.cassandra.db.memtable.ShardBoundaries;
+import org.apache.cassandra.db.memtable.TrieMemtable;
 import org.apache.cassandra.dht.AbstractBounds;
 import org.apache.cassandra.index.sai.IndexContext;
 import org.apache.cassandra.index.sai.plan.Expression;
+import org.apache.cassandra.index.sai.utils.PrimaryKey;
 import org.apache.cassandra.index.sai.utils.PrimaryKeys;
+import org.apache.cassandra.index.sai.utils.RangeConcatIterator;
 import org.apache.cassandra.index.sai.utils.RangeIterator;
+import org.apache.cassandra.utils.MergeIterator;
 import org.apache.cassandra.utils.Pair;
 import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.concurrent.OpOrder;
+
+import static java.util.function.Function.identity;
 
 public class MemtableIndex
 {
-    private final MemoryIndex index;
+    private final ShardBoundaries boundaries;
+    private final MemoryIndex[] rangeIndexes;
+    private final AbstractType<?> validator;
+    private final ClusteringComparator clusteringComparator;
     private final LongAdder writeCount = new LongAdder();
-    private final LongAdder estimatedMemoryUsed = new LongAdder();
+    private final LongAdder estimatedOnHeapMemoryUsed = new LongAdder();
+    private final LongAdder estimatedOffHeapMemoryUsed = new LongAdder();
 
     public MemtableIndex(IndexContext indexContext)
     {
-        this.index = new TrieMemoryIndex(indexContext);
+        this.boundaries = indexContext.owner().localRangeSplits(TrieMemtable.SHARD_COUNT);
```

Comment:
Agreed. I've added an assertion to the `IndexContext.owner` method. I would prefer it if we could get rid of the other `IndexContext` constructors so that this wasn't needed but that is beyond the scope of this ticket because it would need some serious changes to `QueryController` and it's dependent classes.

Reply:
I've changed this so we only have a single `IndexContext` constructor where the `owner` is marked as `@NonNull`.